storage/tests/python/src/norcow.py

```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
 
     def _dump(self):
         return [bytes(sector) for sector in self.sectors]
+
+
+class NorcowBitwise(Norcow):
+    def __init__(self):
+        super().__init__()
+        self.word_size = consts.WORD_SIZE
+        self.magic = consts.NORCOW_MAGIC_AND_VERSION
+        self.item_prefix_len = 4
+        self.lib_name = "libtrezor-storage.so"
+
+    def get_lib_name(self):
+        return self.lib_name
+
+    def is_byte_access(self):
+        return True
+
+    def _is_updatable(self, key: int, new: bytes) -> bool:
+        """
+        Item is updatable if the new value is the same or
+        it changes 1 to 0 only (the flash memory does not
+        allow to flip 0 to 1 unless you wipe it).
+        """
+
+        old, _ = self._find_item(key)
+        if old is None:
+            return False
+        if len(old) != len(new):
+            return False
+        for a, b in zip(old, new):
+            if a & b != b:
+                return False
+        return True
+
+    def _write(self, pos: int, key: int, new_value: bytes) -> int:
+        data = pack("<HH", key, len(new_value)) + align_data(new_value, self.word_size)
+        if pos + len(data) > consts.NORCOW_SECTOR_SIZE:
+            raise RuntimeError("Norcow: item too big")
+        self.sectors[self.active_sector][pos : pos + len(data)] = data
+        return len(data)
+
+    def _norcow_item_length(self, data_len: int) -> int:
+        # APP_ID, KEY_ID, LENGTH, DATA, ALIGNMENT
+        return self.item_prefix_len + data_len + align_int(data_len, self.word_size)
+
+    def _read_item(self, offset: int) -> (int, bytes):
+        if offset >= consts.NORCOW_SECTOR_SIZE:
+            raise ValueError("Norcow: no data on this offset")
+
+        key = self.sectors[self.active_sector][offset : offset + 2]
+        key = int.from_bytes(key, sys.byteorder)
```

Comment:
The following issue wasn't introduced in this pull request but we may want to fix it.

In `_write`, little endian is used, while in `_read_item`, the endianness is retrieved from the system. This is inconsistent and causes that the tests won't work on big-endinan machines. We should either:
  * always use little-endian or
  * always use big-endian or
  * always retrieve the endianness from the system.
 
 By the way, we can use `key, length = struct.unpack('<HH', data)`.

Reply:
given that on the target MCUs we always have little endian and in tests we compare the memory bytewise, i went for _always use little-endian_ option. 

https://github.com/trezor/trezor-firmware/commit/106525346632493c313ccc068158ccc5bc0cb709

but this will likely still not work on big-endian machines, since the C code itself does not enforce little-endian and rely on underlying architecture. I don't think we should complicate the C code to make it enforce little endian, and if we don't do that, then failing the tests on big-endian machines might even be better, because when using sys.byteorder it would actually test data structure that is different then on target MCUs.

i also think that running these tests on big-endian machines is quite unlikely

Reply:
> but this will likely still not work on big-endian machines, since the C code itself does not enforce little-endian and rely on underlying architecture

True, I didn't realize that.

It would be probably most consistent to replace all `sys.byteorder` with `'little'` and assert that `sys.byteorder == 'little'`, but it doesn't have to be done in this pull request.